sky/backends/cloud_vm_ray_backend.py

```diff
@@ -1659,30 +1659,60 @@ def _update_cluster_region(self):
             self.launched_resources = self.launched_resources.copy(
                 region=region)
 
-        @property
-        def cluster_yaml(self):
-            return os.path.expanduser(self._cluster_yaml)
+        def _update_stable_cluster_ips(self,
+                                       max_attempts: int = 1) -> List[str]:
+            cluster_external_ips = backend_utils.get_node_ips(
+                self.cluster_yaml,
+                self.launched_nodes,
+                handle=self,
+                head_ip_max_attempts=max_attempts,
+                worker_ip_max_attempts=max_attempts,
+                get_internal_ips=False)
+            cluster_internal_ips = backend_utils.get_node_ips(
+                self.cluster_yaml,
+                self.launched_nodes,
+                handle=self,
+                head_ip_max_attempts=max_attempts,
+                worker_ip_max_attempts=max_attempts,
+                get_internal_ips=True)
+
+            internal_external_ips = list(
+                zip(cluster_internal_ips, cluster_external_ips))
```

Comment:
Yes, the only requirement we have here is that the order of internal IPs matches that of external IPs. 

Looking at https://github.com/ray-project/ray/blob/30a09bd66c46f29e1463287982bd03afc7ecda19/python/ray/autoscaler/_private/commands.py#L1305, this is the case

Reply:
It seems to depend on the order the following function returns each time it is called:
https://github.com/ray-project/ray/blob/30a09bd66c46f29e1463287982bd03afc7ecda19/python/ray/autoscaler/_private/commands.py#L1315
And the above function depends on the `ec2` function for filtering the nodes by the tag:
https://github.com/skypilot-org/skypilot/blob/235d9d65f5e269a10cbd7bec848a89d830d1d8a0/sky/skylet/providers/aws/node_provider.py#L157-L158
Could we find the doc for this function to make sure the order is deterministic for multiple call?

Reply:
I see - I could not find any guarantee that this function is deterministic for multiple calls. There is also the case of the providers for GCP and Azure, which both call their own respective libraries. I tried looking around and I didn't see a good way of doing this without using Ray's private APIs. What do you think?

Perhaps we could submit a PR or patch `get_worker_node_ips` so that the nodes are sorted?

Reply:
Yea, that could be future work. How about we try to run it multiple times and see if the external_ips and internal_ips always match each other? If that works, probably it is fine for now.